tests/compiler/std/string.ts

```diff
@@ -796,6 +791,34 @@ assert(dtoa(1.1e+128) == "1.1e+128");
 assert(dtoa(1.1e-64) == "1.1e-64");
 assert(dtoa(0.000035689) == "0.000035689");
 
+// The following demonstrates behavior that may seem surprising, but is consistent with the
+// IEEE 754 standard and behaves the same way in other languages that have 32-bit floating point numbers.
+assert(dtoa(<u64>4294967272) == "4294967272.0"); // u64 -> f64, no loss of precision
+assert(dtoa(<u32>4294967272) == "4294967300.0"); // u32 -> f32, loss of precision is expected
```

Comment:
I'm not sure I like how `dtoa<u32>(...)` is treated as if an `f32` was passed in...maybe you should do `if (!isFloat<T>()) ERROR("dtoa only supports floats")` or something? Then again, `dtoa` is internal to AS's stdlib...

Reply:
I have a commented assert for that reason - I forgot to ask.  I wasn't sure what the best practice is.  Is it too much overhead to put it in?

Reply:
Or is it actually better that it casts implicitly?  I mean, `dtoa(123)` worked previously, so if I assert then wouldn't that be a breaking change?

Reply:
Regarding the assert, you should use `if (!isFloat<T>()) ERROR("some error message")`. `ERROR` is the compile-time equivalent of `assert(false)`.

`dtoa(123)` probably would've worked before by upcasting to an `f64`...hm, I guess we can't just bail on integers...you can call `dtoa` recursively with the appropriate float type then. You'd still have to make sure `T` is a float or integer (`T extends number` doesn't do anything in AS).

Reply:
Something like:
```ts
if (isInteger<T>()) return sizeof<T>() < 4 ? dtoa<f32>(value) : dtoa<f64>(value);
```

Reply:
Ok.  How about this?

```ts
export function dtoa<T extends number>(value: T): String {
  if (isInteger<T>()) {
    return dtoa_impl(<f64>value, false);
  }

  if (isFloat<T>()) {
    if (sizeof<T>() == 4) {
      return dtoa_impl(<f32>value, true);
    } else {
      return dtoa_impl(<f64>value, false);
    }
  }
  
  ERROR("dtoa: unsupported type");
}
```

It actually helps with the existing integer case to avoid surprising results.  I don't think it adds much overhead, since the routine is 64-bit anyway, and there's no strange decimals added like in the f32 case.

Reply:
applied that, and similar for `dtoa_buffered`.  Updated the test case.

LMK if you want anything else. 😃

Reply:
Oh, I might've made a mistake...`dtoa_impl` only accepts `f64` so `ERROR` isn't actually necessary...sorry about that.

```ts
export function dtoa<T extends number>(value: T): string {
    const isSingle = isFloat<T>() && sizeof<T>() == 4;
    return dtoa_impl(value, isSingle);
}
```

This might work?

Reply:
That's brilliant!  And the compiler catches it.  Say I try to pass a string, I get:

```
ERROR TS2322: Type '~lib/string/String' is not assignable to type 'f64'.
     :
 746 │ return dtoa_impl(value, isSingle);
     │                  ~~~~~
     └─ in ~lib/util/number.ts(746,20)

FAILURE 1 compile error(s)
Build failed.
```

The location is off, but the error is correct.  And TypeScript editor tooling catches it also.
![image](https://github.com/user-attachments/assets/ada99820-a25e-496a-822c-84184cea7f0f)

Reply:
Applied that.  It reduced some debug wat blocks. 👍